### DISCUSSION TITLE
JSGF and Live Mode

### DISCUSSION THREAD

**Discussion by Kiloku** [cmusphinx/pocketsphinx#369] [Q&A] [answered]
Does `-jsgf` not work with "live" mode? I've tried the `live_pulseaudio` example and it seems to not use my grammar file properly, and ends up with inaccurate recognition. 
When I run `pocketsphinx single` with my grammar file and a .wav I recorded with the same words I said in live mode, it recognizes perfectly.

Here's the output of `pocketsphinx single -jsgf commands.jsgf  alphatwo.wav` (formatting modified by hand for readability)
```JSON
{
    "b": 0.000, "d": 1.220, "p": 0.305, "t": "alpha two",
    "w": [
        { "b": 0.000, "d": 0.010, "p": 1.000, "t": "<s>" },
        { "b": 0.000, "d": 0.190, "p": 0.333, "t": "<sil>" },
        { "b": 0.190, "d": 0.380, "p": 0.961, "t": "alpha" },
        { "b": 0.570, "d": 0.650, "p": 1.000, "t": "two" },
        { "b": 1.210, "d": 0.010, "p": 1.000, "t": "</s>" }
    ]
}
```

Now here's the output of `live_pulseaudio -jsgf commands.jsgf` (Note: At the end it converted into a rather impolite phrase :sweat_smile: )
```Speech start at 1.08
PARTIAL RESULT: oh
PARTIAL RESULT: oh
PARTIAL RESULT: ill
PARTIAL RESULT: ill
PARTIAL RESULT: elf
PARTIAL RESULT: elf
PARTIAL RESULT: elf
PARTIAL RESULT: ill for
PARTIAL RESULT: alfred
PARTIAL RESULT: elf it
PARTIAL RESULT: elf it
PARTIAL RESULT: elf it
PARTIAL RESULT: elf it
PARTIAL RESULT: elf it
PARTIAL RESULT: elf it she
PARTIAL RESULT: elf it you
PARTIAL RESULT: elf it you
PARTIAL RESULT: elf it you
PARTIAL RESULT: elf it you
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
PARTIAL RESULT: elf it to
Speech end at 2.28
oh fuck you
```
None of the words it recognized in live mode are in the JSGF grammar I wrote, which is as follows:

```
#JSGF V1.0 ISO8859-1 en;

grammar scp.khonsu.commands;

public <targetedCommand> = <action> <target>;
public <call> = <wingName> | <fighterName>;
public <singleCommand> = form on my wing |  ((cover | protect) me) | 
                  (engage | attack) [the | all] (enemy | enemies) | depart;

<action> = destroy | attack | disable | disarm | protect | ignore;
<target> = [that] (target | subsystem) | [my] (target | attacker);


<greekLetter> = alpha | beta | gamma | delta | epsilon | zeta | theta | iota | kappa | lambda; 

<number> = one | two | three | four | five | six | seven | eight | nine | ten;

<wingName> = <greekLetter> [wing] | all (fighters | wings);

<fighterName> = <greekLetter> <number>; 
``` 

Edit to add: I tried `pocketsphinx live` instead of the `live_pulseaudio` example, but it didn't recognize anything at all, which I believe is because it didn't know how to access my audio input without using pulseaudio.

**Comment by Kiloku** [comment, marked as answer]
I figured it out. I didn't realize, but the `live_pulseaudio` example does not use the config parameters at all! 
When I manually edited the source to set the configs directly to use my JSGF file, it worked as expected. 

I think I can work further from that now.

**Comment by dhdaines** [comment]
Hi! Unfortunately I was away for the weekend before I could answer, but I'm glad you figured it out.  As you discovered, `live_pulseaudio` doesn't do any command-line parsing at all, so if you want to set some configuration options you have to do it in the code, in the call to `ps_config_init` for example.
